### PR TITLE
LUN-39: Logic for exercise 2

### PR DESCRIPTION
**Pull request by SteffiStuffel**
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

### CODE REVIEW

**Review by klinzo — CHANGES_REQUESTED** [review]
I am getting the following warning:
`warn Package @types/styled-components-react-native has been ignored because it contains invalid configuration. Reason: Cannot find module '@types/styled-components-react-native/package.json'`
Seems like this just adds `Artikel üben` to the available exercises which looks like exercise three (three trophies), am I wrong here?

Other options are clickable after selecting the right one, guess that should not be possible. Working fine apart from that already, good work!

**Comment by SteffiStuffel** [comment]
> I am getting the following warning:
> `warn Package @types/styled-components-react-native has been ignored because it contains invalid configuration. Reason: Cannot find module '@types/styled-components-react-native/package.json'`

I do not get this warning.

> Seems like this just adds `Artikel üben` to the available exercises which looks like exercise three (three trophies), am I wrong here?

No you are right, i messed this up. I think we should use names for the exercises. I suggest: Single Choice Exercise, Article Choice Exercise and Writing Exercise. What do you think?

> Other options are clickable after selecting the right one, guess that should not be possible. Working fine apart from that already, good work!

Good catch. Fixed it.

File: src/components/SingleChoiceListItem.tsx [hunk 1]
```diff
@@ -15,17 +15,17 @@ const StyledContainer = styled.View`
   justify-content: flex-start;
   flex-direction: row;
   align-items: baseline;
-  border-color: ${(props: { pressed: boolean; selected: boolean }) => {
+  border-color: ${props => {
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/components/SingleChoiceListItem.tsx:18]: Yes, it makes auto types. But anyway the types defined here are ignored.

E.g. props.pressed is boolean, but 
` background-color: ${(props: { pressed: string; ... }) => {
    console.log(typeof props.pressed)`
outputs boolean. 

So I think this should be removed everywhere.

File: src/components/SingleChoiceListItem.tsx [hunk 2]
```diff
@@ -15,17 +15,17 @@ const StyledContainer = styled.View`
...
     if (props.pressed || props.selected) {
       return 'transparent'
     } else {
       return COLORS.lunesBlackUltralight
     }
   }};
-  background-color: ${(props: { pressed: boolean; selected: boolean; correct: boolean }) => {
+  background-color: ${(props: { pressed: boolean; selected: boolean; correct: boolean; onPress: any }) => {
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/components/SingleChoiceListItem.tsx:25]: Removing the types completely defaults in ` type 'ThemedStyledProps  , DefaultTheme>'`, `any` would work, but it's not nice, so I changed it to `SingleChoiceListItemType`.

**Comment by klinzo** [comment]
> > I am getting the following warning:
> > `warn Package @types/styled-components-react-native has been ignored because it contains invalid configuration. Reason: Cannot find module '@types/styled-components-react-native/package.json'`
> 
> I do not get this warning.

Probably I just need to run `yarn` :grin: 

> 
> > Seems like this just adds `Artikel üben` to the available exercises which looks like exercise three (three trophies), am I wrong here?
> 
> No you are right, i messed this up. I think we should use names for the exercises. I suggest: Single Choice Exercise, Article Choice Exercise and Writing Exercise. What do you think?

I like that :+1:

File: src/components/SingleChoiceListItem.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/components/SingleChoiceListItem.tsx:18]: Oke, removing is not working. I added the types as before. Maybe we can make a deeper research, if there is a better handling than adding types manually.

File: src/components/SingleChoiceListItem.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/components/SingleChoiceListItem.tsx:25]: done.

File: src/constants/data.ts [hunk 3]
```diff
@@ -65,10 +65,13 @@ export const ARTICLES = {
   die: 'die',
   der: 'der',
   das: 'das',
-  diePlural: 'die (plural)'
-}
+  diePlural: 'Die'
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/constants/data.ts:68]: yes. Haven't seen a better solution.

File: src/screens/InitialSummaryScreen.tsx [hunk 4]
```diff
@@ -68,13 +68,18 @@ const InitialSummaryScreen = ({ navigation, route }: InitialSummaryScreenPropsTy
...
   useFocusEffect(
     React.useCallback(() => {
-      AsyncStorage.getExercise(exercise)
-        .then(value => {
-          if (value !== null) {
-            setResults(Object.values(value[disciplineTitle][trainingSet]))
-          }
-        })
-        .catch(e => console.error(e))
+      if (exercise === 3) {
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/screens/InitialSummaryScreen.tsx:71]: Only exercise 4 (numbered 3) - the writing exercise - is stored in the AsyncStorage. For exercise 2 & 3 (single choice and article choice) the results are passed with the extraParams to to result screen.

I think AsyncStorage is not needed, as the redux store should really be enough to keep the state of the exercise. Maybe we can discuss this today in the evening? What do you think?

File: src/screens/LearnArticlesExerciseScreen.tsx [hunk 5]
```diff
@@ -1,8 +1,126 @@
...
+interface LearnArticlesExerciseScreenPropsType {
+  route: RouteProp<RoutesParamsType, 'VocabularyOverview'>
+  navigation: StackNavigationProp<RoutesParamsType, 'VocabularyOverview'>
+}
+
+const LearnArticlesExerciseScreen = ({ navigation, route }: LearnArticlesExerciseScreenPropsType) => {
+  const { trainingSetId } = route.params.extraParams
+  const [documents, setDocuments] = useState<DocumentsType>([])
+  const [isLoading, setIsLoading] = useState<boolean>(true)
+  const [isFinished, setIsFinished] = useState<boolean>(false)
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/screens/LearnArticlesExerciseScreen.tsx:30]: isLoading is about fetching the data for the exercise.
isFinished changes if the user has selected an answer.

Naming is used also in the writing-exercise and renaming must be done in lots of places. Pleas open a ticket for it, if you think it is really necessary, as it is not part of this ticket, as i just continued the existing naming.

File: src/screens/LearnArticlesExerciseScreen.tsx [hunk 6]
```diff
@@ -1,8 +1,126 @@
...
+      const result: DocumentResultType = { ...documents[currentWord], result: SIMPLE_RESULTS.incorrect }
+      setResults([...results, result])
+    }
+    setAnswerOptions(answerOptionsUpdated)
+    setIsFinished(true)
+  }
+
+  const buttonClick = () => {
+    const exersiceFinished = currentWord + 1 >= count
+    const extraParamsWithResults: any = extraParams
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/screens/LearnArticlesExerciseScreen.tsx:95]: done

File: src/screens/SingleChoiceExerciseScreen.tsx [hunk 7]
```diff
@@ -1,45 +1,68 @@
...
+interface SingleChoiceExerciseScreenPropsType {
+  route: RouteProp<RoutesParamsType, 'VocabularyOverview'>
+  navigation: StackNavigationProp<RoutesParamsType, 'VocabularyOverview'>
+}
+
+const SingleChoiceExerciseScreen = ({ navigation, route }: SingleChoiceExerciseScreenPropsType) => {
+  const { trainingSetId } = route.params.extraParams
+  const [documents, setDocuments] = useState<DocumentsType>([])
+  const [isLoading, setIsLoading] = useState<boolean>(true)
+  const [isFinished, setIsFinished] = useState<boolean>(false)
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/screens/SingleChoiceExerciseScreen.tsx:24]: won't do, see comment above.

File: src/screens/SingleChoiceExerciseScreen.tsx [hunk 8]
```diff
@@ -1,45 +1,68 @@
...
+  const [count, setCount] = useState<number>(0)
+  const [answerOptions, setAnswerOptions] = useState<SingleChoiceListItemType[]>([])
+  const [currentWord, setCurrentWord] = useState<number>(0)
+
+  useEffect(() => {
+    const url = ENDPOINTS.documents.all.replace(':id', `${trainingSetId}`)
+    axios.get(url).then(response => {
+      console.log('get', trainingSetId)
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/screens/SingleChoiceExerciseScreen.tsx:32]: done

File: src/screens/SingleChoiceExerciseScreen.tsx [hunk 9]
```diff
@@ -1,45 +1,68 @@
...
+  navigation: StackNavigationProp<RoutesParamsType, 'VocabularyOverview'>
+}
+
+const SingleChoiceExerciseScreen = ({ navigation, route }: SingleChoiceExerciseScreenPropsType) => {
+  const { trainingSetId } = route.params.extraParams
+  const [documents, setDocuments] = useState<DocumentsType>([])
+  const [isLoading, setIsLoading] = useState<boolean>(true)
+  const [isFinished, setIsFinished] = useState<boolean>(false)
+  const [count, setCount] = useState<number>(0)
+  const [answerOptions, setAnswerOptions] = useState<SingleChoiceListItemType[]>([])
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/screens/SingleChoiceExerciseScreen.tsx:26]: I'm a bit confused here. 

If I remove answerOptions from state, i would have them as variable. Problem: I would have to add them to the dependency array of buildAnswerOption(), but i do not want buildAnswerOption() to be executed, after an answer is selected and the answerOptions get updated. 
Also when selecting an answer and therefore answerOptions change, we want to have a rerender.

File: src/screens/SingleChoiceExerciseScreen.tsx [hunk 10]
```diff
@@ -1,45 +1,68 @@
...
+    const url = ENDPOINTS.documents.all.replace(':id', `${trainingSetId}`)
+    axios.get(url).then(response => {
+      console.log('get', trainingSetId)
+      setDocuments(response.data)
+      setCount(response.data.length)
+      setIsLoading(false)
+    })
+  }, [trainingSetId])
+
+  const buildAnswerOption = React.useCallback(() => {}, [currentWord, documents])
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/screens/SingleChoiceExerciseScreen.tsx:39]: Will insert the correct code for building the answerOptions in LUN-40. Already doing this.

**Comment by SteffiStuffel** [comment]
> > No you are right, i messed this up. I think we should use names for the exercises. I suggest: Single Choice Exercise, Article Choice Exercise and Writing Exercise. What do you think?
> 
> I like that

opened a ticket for it.

File: src/components/SingleChoiceListItem.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply src/components/SingleChoiceListItem.tsx:17]: I think Sarah would now more about this topic. There is also this issue: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/53490

File: src/screens/InitialSummaryScreen.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply src/screens/InitialSummaryScreen.tsx:71]: Ahh okay, I was wrong with the numbering and thought this was exercise 3 and not 4. Could you replace this with a constant? I agree that no async storage is needed then probably, saving anything if users abort the exercise is probably op for this.

File: src/screens/LearnArticlesExerciseScreen.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply src/screens/LearnArticlesExerciseScreen.tsx:30]: This is new naming in this components and to me it is very unclear what this var is about on first reading. Please change it for your new components already :)

File: src/screens/SingleChoiceExerciseScreen.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply src/screens/SingleChoiceExerciseScreen.tsx:24]: Please do, this is new code :see_no_evil:

File: src/screens/SingleChoiceExerciseScreen.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply src/screens/SingleChoiceExerciseScreen.tsx:26]: lets discuss this later :)

File: src/screens/InitialSummaryScreen.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/screens/InitialSummaryScreen.tsx:71]: Done.

**Review by klinzo — APPROVED** [review]

**Pull request merged**